group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -4209,31 +4241,67 @@ private void removePendingSyncMember(
      * @param context        The request context.
      * @param request        The actual Heartbeat request.
      *
-     * @return The Heartbeat response.
+     * @return The coordinator result that contains the heartbeat response.
      */
-    public HeartbeatResponseData classicGroupHeartbeat(
+    public CoordinatorResult<HeartbeatResponseData, CoordinatorRecord> classicGroupHeartbeat(
```

Comment:
maybe i'm missing something but i don't see where we actually initialize CoordinatorResult with records to write to the log

Reply:
We don't generate any records. However, as we access the uncommitted state, we must use a write operation to ensure that we don't return before the state read is committed.

Reply:
so this is similar to the offset fetch path. how come we want to access the uncommitted state now?

Reply:
I guess because now we'll also read from the ConsumerGroup related things which are timeline data structures.

Reply:
Yeah. This is correct. With the classic group type, the group state is not based on timeline data structures so using a read operation was fine.